Update to 1.12.2

ExpensiveKoala opened the pull request.

ExpensiveKoala committed:
Fixed Most compile errors

Still need to...
convert all recipes to json,
fix block/item registering,
fix all deprecated stuff

ExpensiveKoala committed:
Add registry events

ExpensiveKoala committed:
Finish ContentModule

Now just need to fix all of the modules that extend it.

ExpensiveKoala committed:
Finish moving block/item registry to events.

ExpensiveKoala committed:
Fix library & Crafttweaker

ExpensiveKoala committed:
Move model registry to event

ExpensiveKoala committed:
Move item color over to postInit

ExpensiveKoala committed:
Added recipe examples.

To add a condition/config option, add to ConfigConditionFactory

elifoster commented:
The build failed because you imported a `main` class (`Config`) into an `api` class (`ConfigConditionFactory`). That dependency is the other way around:
* api
  * main
    * test

ExpensiveKoala committed:
Move ConditionFactory to main

ExpensiveKoala committed:
Don't use FMLLog

elifoster commented on build.gradle:
```diff
@@ -116,15 +116,36 @@ repositories {
+		url = "http://maven.blamejared.com/"
+	}
+	
+	//Baubles
```
Style change: `//Baubles` to `// Baubles`

elifoster commented on build.gradle:
```diff
@@ -116,15 +116,36 @@ repositories {
     compile naturaVersion
-    compile ic2Version
-    compile jeiVersion
+    //compile ic2Version
```
Style change: `//compile ic2Version` to `// compile ic2Version`

elifoster commented on gradle.properties:
```diff
@@ -8,7 +8,13 @@ mantleVersion=slimeknights.mantle:Mantle:1.12-1.3.1.22
+enchiridionVersion=uk.joshiejack.enchiridion:enchiridion_1.12.2:1.12.2-3.3.3-4
+craftTweaker2APIVersion=CraftTweaker2:CraftTweaker2-API:4.1.0+
+craftTweaker2Version=CraftTweaker2:CraftTweaker2-MC1120-Main:1.12-4.1.0+
+#This is on the commit number. NOT the version.
```
Style change: `#This is on the commit number. NOT the version.` to `# This is on the commit number. NOT the version.`

elifoster commented on src/api/java/eiteam/esteemedinnovation/api/book/BookPageCrafting.java:
```diff
@@ -49,6 +53,43 @@ public BookPageCrafting(String name, String... keys) {
+                }
+            }
+        }
+        //<editor-fold desc="Old Code">
```
Old code should be deleted instead of commented out

elifoster commented on src/api/java/eiteam/esteemedinnovation/api/book/BookPageCrafting.java:
```diff
@@ -49,6 +53,43 @@ public BookPageCrafting(String name, String... keys) {
+                        if (inputs[i] instanceof Collection) {
+                            newList.addAll((Collection) inputs[i]);
+                        } else {
+                            newList.add(inputs[i]); //probably not going to ever happen
```
Comments go above the line, not at the end

elifoster replied:
Why is this probably not going to ever happen?

ExpensiveKoala replied:
Because it only ever sets inputs[i] as a Collection<Object>, so that if statement is always true. I kept it because the old code had it.

elifoster commented on src/api/java/eiteam/esteemedinnovation/api/book/BookPageCrafting.java:
```diff
@@ -49,6 +53,43 @@ public BookPageCrafting(String name, String... keys) {
+                            newList.add(inputs[i]); //probably not going to ever happen
+                        }
+                    }
+                    if (recipe.getIngredients().size() > i && recipe.getIngredients().get(i) != Ingredient.EMPTY) {
```
Is there not an `isEmpty()` method on `Ingredient`?

ExpensiveKoala replied:
There isn't.

elifoster commented on src/api/java/eiteam/esteemedinnovation/api/book/BookPageCrafting.java:
```diff
@@ -49,6 +53,43 @@ public BookPageCrafting(String name, String... keys) {
+                    inputs[i] = newList;
+                }
+
+            } else if (recipe instanceof ShapelessRecipeFactory.ShapelessRecipe) {
```
I feel like we could reduce redundant code here somehow. It is almost exactly the same as the above branch.

ExpensiveKoala replied:
Looking at it, yes we probably can.

elifoster commented on src/api/java/eiteam/esteemedinnovation/api/book/BookPageCrafting.java:
```diff
@@ -121,7 +164,7 @@ public BookPageCrafting(String name, IRecipe... recipes) {
             recipes.add(BookRecipeRegistry.getRecipe(key));
         }
-        return recipes.toArray(new IRecipe[recipes.size()]);
+        return recipes.size() > 0 ? recipes.toArray(new IRecipe[recipes.size()]) : new IRecipe[0];
```
The `size()` check should be replaced with `!isEmpty()`

elifoster commented on src/api/java/eiteam/esteemedinnovation/api/recipe/ShapedRecipeFactory.java:
```diff
@@ -0,0 +1,38 @@
+import net.minecraftforge.common.crafting.JsonContext;
+import net.minecraftforge.oredict.ShapedOreRecipe;
+
+public class ShapedRecipeFactory implements IRecipeFactory{
```
Missing space before `{`

elifoster replied:
Can you provide documentation here in a JavaDoc?

elifoster commented on src/api/java/eiteam/esteemedinnovation/api/recipe/ShapedRecipeFactory.java:
```diff
@@ -0,0 +1,38 @@
+        return shapedRecipe;
+    }
+
+    public static class ShapedRecipe extends ShapedOreRecipe {
```
Is this necessary? It is exactly the same thing as `ShapedOreRecipe`.

ExpensiveKoala replied:
It wouldn't be necessary if we don't need the "key" used to sort/store the recipes in the book registry.
This allows the json file to have the book key in the recipe to match the old method of adding recipes to the `BookRecipeRegistry`.

elifoster commented on src/api/java/eiteam/esteemedinnovation/api/recipe/ShapelessRecipeFactory.java:
```diff
@@ -0,0 +1,43 @@
+import net.minecraftforge.common.crafting.JsonContext;
+import net.minecraftforge.oredict.ShapelessOreRecipe;
+
+public class ShapelessRecipeFactory implements IRecipeFactory{
```
Missing space before `{`

elifoster replied:
Can you provide documentation here in a JavaDoc?

elifoster commented on src/api/java/eiteam/esteemedinnovation/api/recipe/ShapelessRecipeFactory.java:
```diff
@@ -0,0 +1,43 @@
+        return shapelessRecipe;
+    }
+
+    public static class ShapelessRecipe extends ShapelessOreRecipe {
```
s this necessary? It is exactly the same thing as `ShapelessOreRecipe`.

ExpensiveKoala replied:
See `ShapedRecipe`

elifoster commented on src/api/java/eiteam/esteemedinnovation/api/recipe/ShapelessRecipeFactory.java:
```diff
@@ -0,0 +1,43 @@
+        ItemStack itemstack = CraftingHelper.getItemStack(JsonUtils.getJsonObject(json, "result"), context);
+
+        ShapelessRecipe shapelessRecipe = new ShapelessRecipe(group.isEmpty() ? null : new ResourceLocation(group), ings, itemstack);
+        BookRecipeRegistry.addRecipe(JsonUtils.getString(json, "book_key"), shapelessRecipe);
```
Because this is used here and in `ShapedRecipeFactory` also, it might be a good idea to move this `"book_key"` String into a constant. What do you think?

ExpensiveKoala replied:
Might be a good idea.

elifoster commented on src/api/java/eiteam/esteemedinnovation/api/recipe/ShapelessRecipeFactory.java:
```diff
@@ -0,0 +1,43 @@
+        String group = JsonUtils.getString(json, "group", "");
+
+        NonNullList<Ingredient> ings = NonNullList.create();
+        for (JsonElement ele : JsonUtils.getJsonArray(json, "ingredients"))
```
Missing braces {} around for loop body

elifoster commented on src/api/java/eiteam/esteemedinnovation/api/recipe/ShapelessRecipeFactory.java:
```diff
@@ -0,0 +1,43 @@
+        for (JsonElement ele : JsonUtils.getJsonArray(json, "ingredients"))
+            ings.add(CraftingHelper.getIngredient(ele, context));
+
+        if (ings.isEmpty())
```
Missing braces {} around if body.

elifoster commented on src/api/java/eiteam/esteemedinnovation/api/recipe/ShapelessRecipeFactory.java:
```diff
@@ -0,0 +1,43 @@
+
+public class ShapelessRecipeFactory implements IRecipeFactory{
+    @Override
+    public IRecipe parse(JsonContext context, JsonObject json) {
```
If I'm not mistaken, shouldn't there be a throws clause in this function declaration?

ExpensiveKoala replied:
Maybe. The code could throw a `JsonSyntaxException`, but I haven't seen anyone add a throws clause here.

elifoster commented on src/main/java/eiteam/esteemedinnovation/commons/init/ContentModule.java:
```diff
@@ -44,16 +68,18 @@ public void create(Side side) {}
     /**
-     * Handle any client-side preInit stuff here. Most likely, you'll want to use this to register your models. Called
+     * Handle any client-side preInit stuff here. Called
      * in preInit on the client after {@link #oreDict(Side)}.
```
Unnecessary line break

elifoster commented on src/main/java/eiteam/esteemedinnovation/commons/init/ContentModule.java:
```diff
@@ -33,9 +34,32 @@
-     * Initialize and register your items. Called during preInit in both client and server.
-     * Use the setup methods below to easily handle registration; or just handle everything on your own.
+     * Called during init in both client and server.
+     * Use to register miscellaneous things for the module.
```
Unnecessary line break

elifoster commented on src/main/java/eiteam/esteemedinnovation/commons/init/ContentModule.java:
```diff
@@ -79,10 +105,10 @@ public void postInitClient() {}
 
     /**
-     * Overload for {@link #setup(Item, String, CreativeTabs)} that uses {@link EsteemedInnovation#tab} for the tab argument.
+     * Overload for {@Link #setup(RegistryEvent.Register, Item, String, CreativeTabs)} that uses {@link EsteemedInnovation#tab} for the tab argument.
```
`@Link` -> `@link`

elifoster commented on src/main/java/eiteam/esteemedinnovation/commons/init/ContentModule.java:
```diff
@@ -93,66 +119,67 @@ protected Item setup(Item startingItem, String path) {
      * @return The registered block (ItemBlock is not returned).
      */
-    protected Block setup(Block startingBlock, String path, CreativeTabs tab, Function<Block, ItemBlock> itemBlockFunc) {
+    protected Block setup(RegistryEvent.Register<Block> event, Block startingBlock, String path, CreativeTabs tab) {
```
Do we no longer have to worry about `ItemBlock`s? It looks like all the stuff relating to `ItemBlock`s was removed. if so, the `@return` in the docs should be updated.

ExpensiveKoala replied:
The setup for Blocks can only register blocks so we can't do anything with `ItemBlock`s here.

elifoster commented on src/main/java/eiteam/esteemedinnovation/commons/init/ContentModule.java:
```diff
@@ -93,66 +119,67 @@ protected Item setup(Item startingItem, String path) {
+    }
+
+
+    protected void setupItemBlock(RegistryEvent.Register<Item> event, Block startingBlock, CreativeTabs tab, Function<Block, ItemBlock> itemBlock) {
```
Why do we need separate functions for handing `ItemBlock`s? That seems like unnecessary work.

ExpensiveKoala replied:
That is there just to make registering `ItemBlock`s easier. They are also different from Items in the sense that they don't need the registry name (It'll take from the `Block`) and they also need the block they represent. So these are here mainly for helping register the `ItemBlock`s.

elifoster replied:
Hm. It might be an idea to have a `Map` of `Block => Function<Block, ItemBlock>` that is added to in `registerBlock` so we can still automatically register `ItemBlock`s without having to manually do it. What do you think?

ExpensiveKoala replied:
Hmm. I guess, but the way I can see the implementation of that would only move code that you have in the `registerItems` method into the creation of that map. It wouldn't really change the amount of work required.

elifoster commented on src/main/java/eiteam/esteemedinnovation/commons/init/ContentModuleHandler.java:
```diff
@@ -72,7 +77,7 @@ public static void preInit() {
         for (ContentModule module : modules) {
             module.create(side);
-            module.oreDict(side);
+            //module.oreDict(side);
```
Uncomment this. It's deprecated but we can still call it for the time being so it's easier to locate where this function is used.

elifoster commented on src/main/java/eiteam/esteemedinnovation/commons/init/ContentModuleHandler.java:
```diff
@@ -83,7 +88,7 @@ public static void init() {
         boolean isClient = side == Side.CLIENT;
         for (ContentModule module : modules) {
-            module.recipes(side);
+            //module.recipes(side);
```
Uncomment this. It's deprecated but we can still call it for the time being so it's easier to locate where this function is used.

elifoster commented on src/main/java/eiteam/esteemedinnovation/commons/init/ContentModuleHandler.java:
```diff
@@ -101,6 +106,25 @@ public static void postInit() {
+    }
+
+    @SideOnly(Side.CLIENT)
+    public static void registerModels(ModelRegistryEvent event) {
```
Would it be possible to generify these methods?

ExpensiveKoala replied:
Maybe, but forge fires `RegistryEvent`s for the different ["registries"](https://mcforge.readthedocs.io/en/latest/concepts/registries/#registering-things).

elifoster commented on src/main/java/eiteam/esteemedinnovation/commons/recipe/ConfigConditionFactory.java:
```diff
@@ -0,0 +1,24 @@
+import java.util.function.BooleanSupplier;
+
+public class ConfigConditionFactory implements IConditionFactory {
+
```
Unnecessary line break

elifoster commented on src/main/java/eiteam/esteemedinnovation/commons/recipe/ConfigConditionFactory.java:
```diff
@@ -0,0 +1,24 @@
+
+        switch (configSetting) {
+            //TODO: add config options here
+            case "enableAstrolabe": return () -> Config.enableAstrolabe;
```
I think if we are going to be overhauling the way configuration works, we should put config options into `ContentModule`s. Do you have any ideas for how we could do that? I've never heard of this `ConditionFactory` stuff.

ExpensiveKoala replied:
Since moving to json recipes, we need a way to enable/disable certain recipes depending on config settings. Forge has some conditions built in for if a certain mod is loaded, etc... `ConfigConditionFactory` allows us to have separate config options to disable each individual recipe. (As the old code had)

elifoster replied:
I will mess around with this after merged. I'm fine with this for the purposes of this PR, will probably split it up into the ContentModules before first release though.

elifoster commented on src/main/java/eiteam/esteemedinnovation/converter/TileEntityFluidSteamConverter.java:
```diff
@@ -39,7 +39,7 @@
         Fluid steam = FluidRegistry.getFluid("steam");
         if (steam == null && CrossMod.IC2) {
-            steam = FluidRegistry.getFluid("ic2steam");
+            //steam = FluidRegistry.getFluid("ic2steam");
```
Why is this commented out?

ExpensiveKoala replied:
IC2's coremod was crashing so I removed it. Kinda commented this out as a precaution that it won't somehow crash here as well. (It shouldn't)

elifoster commented on src/main/java/eiteam/esteemedinnovation/hammer/ContainerSteamAnvil.java:
```diff
@@ -163,7 +163,7 @@ public void updateRepairOutput() {
 
             if (itemstack2 != null) {
-                flag = itemstack2.getItem() == Items.ENCHANTED_BOOK && Items.ENCHANTED_BOOK.getEnchantments(itemstack2).tagCount() > 0;
+                flag = itemstack2.getItem() == Items.ENCHANTED_BOOK && EnchantmentHelper.getEnchantments(itemstack2).size() > 0;
```
Replace this size check with an empty check

elifoster commented on src/main/java/eiteam/esteemedinnovation/materials/MaterialsModule.java:
```diff
@@ -139,7 +148,7 @@ public void oreDict(Side side) {
         OreDictionary.registerOre(NUGGET_BRASS, new ItemStack(METAL_NUGGET, 1, BRASS_NUGGET.getMeta()));
         OreDictionary.registerOre(NUGGET_GILDED_IRON, new ItemStack(METAL_NUGGET, 1, GILDED_IRON_NUGGET.getMeta()));
-        OreDictionary.registerOre(NUGGET_IRON, new ItemStack(METAL_NUGGET, 1, IRON_NUGGET.getMeta()));
+        OreDictionary.registerOre(NUGGET_IRON, new ItemStack(METAL_NUGGET, 1, ItemMetalNugget.Types.IRON_NUGGET.getMeta()));
```
We should probably remove our Iron Nugget altogether, since Vanilla has one.

You don't have to change this, note to myself, but if you did that'd be cool ;)

elifoster commented on src/main/java/eiteam/esteemedinnovation/materials/raw/BlockGenericOre.java:
```diff
@@ -48,10 +49,11 @@ public IBlockState getStateFromMeta(int meta) {
-        for (OreBlockTypes type : OreBlockTypes.LOOKUP) {
-            list.add(new ItemStack(this, 1, type.getMetadata()));
+    public void getSubBlocks(CreativeTabs itemIn, NonNullList<ItemStack> items) {
+        if(itemIn == EsteemedInnovation.tab) {
```
Missing space after `if`

elifoster commented on src/main/java/eiteam/esteemedinnovation/materials/raw/BlockGenericOre.java:
```diff
@@ -48,10 +49,11 @@ public IBlockState getStateFromMeta(int meta) {
-    public void getSubBlocks(@Nonnull Item item, CreativeTabs tabs, NonNullList<ItemStack> list) {
-        for (OreBlockTypes type : OreBlockTypes.LOOKUP) {
-            list.add(new ItemStack(this, 1, type.getMetadata()));
+    public void getSubBlocks(CreativeTabs itemIn, NonNullList<ItemStack> items) {
```
`itemIn` -> `tab`

elifoster commented on src/main/java/eiteam/esteemedinnovation/materials/refined/BlockBeacon.java:
```diff
@@ -33,9 +34,11 @@ public boolean isBeaconBase(IBlockAccess worldObj, BlockPos pos, BlockPos beacon
-    public void getSubBlocks(Item item, CreativeTabs tab, NonNullList<ItemStack> list) {
-        for (MetalBlockTypes type : MetalBlockTypes.values()) {
-            list.add(new ItemStack(this, 1, type.getMetadata()));
+    public void getSubBlocks(CreativeTabs itemIn, NonNullList<ItemStack> items) {
```
`itemIn` -> `tab`

elifoster commented on src/main/java/eiteam/esteemedinnovation/materials/refined/BlockBeacon.java:
```diff
@@ -33,9 +34,11 @@ public boolean isBeaconBase(IBlockAccess worldObj, BlockPos pos, BlockPos beacon
-        for (MetalBlockTypes type : MetalBlockTypes.values()) {
-            list.add(new ItemStack(this, 1, type.getMetadata()));
+    public void getSubBlocks(CreativeTabs itemIn, NonNullList<ItemStack> items) {
+        if(itemIn == EsteemedInnovation.tab) {
```
Missing space after `if`

elifoster commented on src/main/java/eiteam/esteemedinnovation/materials/refined/ItemMetalIngot.java:
```diff
@@ -13,10 +14,11 @@ public ItemMetalIngot() {
-        for (Types type : Types.values()) {
-            subItems.add(new ItemStack(this, 1, type.getMeta()));
+    public void getSubItems(CreativeTabs tab, NonNullList<ItemStack> items) {
+        if(tab == EsteemedInnovation.tab) {
```
Missing space after `if`

elifoster commented on src/main/java/eiteam/esteemedinnovation/materials/refined/ItemMetalNugget.java:
```diff
@@ -13,10 +14,11 @@ public ItemMetalNugget() {
-        for (Types type : Types.values()) {
-            subItems.add(new ItemStack(this, 1, type.getMeta()));
+    public void getSubItems(CreativeTabs tab, NonNullList<ItemStack> items) {
+        if(tab == EsteemedInnovation.tab) {
```
Missing space after `if`

elifoster replied:
Come to think of it, what is the point of this if statement in all of these item classes?

ExpensiveKoala replied:
Starting in 1.12 (I think) unless we have that if statement there, all of the different subitems will end up in every single creative tab.

elifoster commented on src/main/java/eiteam/esteemedinnovation/materials/refined/plates/ItemMetalPlate.java:
```diff
@@ -20,10 +21,11 @@ public ItemMetalPlate() {
-        for (Types type : Types.values()) {
-            subItems.add(new ItemStack(this, 1, type.getMeta()));
+    public void getSubItems(CreativeTabs tab, NonNullList<ItemStack> items) {
+        if(tab == EsteemedInnovation.tab) {
```
See ItemMetalNugget

elifoster commented on src/main/java/eiteam/esteemedinnovation/misc/ItemCraftingComponent.java:
```diff
@@ -15,10 +16,11 @@ public ItemCraftingComponent() {
-        for (Types component : Types.values()) {
-            subItems.add(new ItemStack(MiscellaneousModule.COMPONENT, 1, component.getMetadata()));
+    public void getSubItems(CreativeTabs tab, NonNullList<ItemStack> items) {
+        if(tab == EsteemedInnovation.tab) {
```
See ItemMetalNugget

elifoster commented on src/main/java/eiteam/esteemedinnovation/naturalphilosophy/ItemResearchLog.java:
```diff
@@ -2,27 +2,29 @@
-    @SideOnly(Side.CLIENT)
     @Override
-    public void addInformation(ItemStack stack, EntityPlayer player, List<String> tooltip, boolean advanced) {
+    public void addInformation(ItemStack stack, @Nullable World worldIn, List<String> tooltip, ITooltipFlag flagIn) {
```
Hm. I should probably specify in the code style to not do Forge's annoying `In` parameter suffixes because they annoy the living shit out of me. Not going to require you to change it because it's not an explicitly established style. This is more of a note to myself.

ExpensiveKoala replied:
I hate them too. I just generated the methods without touching the variable names.

elifoster commented on src/main/java/eiteam/esteemedinnovation/smasher/ItemSmashedOre.java:
```diff
@@ -63,10 +64,10 @@ public void addSmelting() {
-    @SideOnly(Side.CLIENT)
-    public void getSubItems(Item item, CreativeTabs tabs, NonNullList<ItemStack> list) {
+    public void getSubItems(CreativeTabs tab, NonNullList<ItemStack> items) {
+        if(tab == EsteemedInnovation.tab)
```
Missing braces and space before if

elifoster commented on src/main/java/eiteam/esteemedinnovation/workshop/SteamWorkshopModule.java:
```diff
@@ -21,9 +22,9 @@ public void recipes(Side side) {
         //noinspection deprecation
         STEAM_ENGINEER_PROFESSION = new VillagerRegistry.VillagerProfession(Constants.EI_MODID + ":steam_engineer",
-          Constants.EI_MODID + ":textures/models/villager.png");
+          Constants.EI_MODID + ":textures/models/villager.png", Constants.EI_MODID + ":textures/models/zombie_villager.png");
```
Can the new parameter be put onto a new line, just for consistency's sake?

elifoster commented on src/main/java/eiteam/esteemedinnovation/armor/exosuit/plates/ItemExosuitPlate.java:
```diff
@@ -24,10 +24,11 @@ public EnumRarity getRarity(ItemStack stack) {
-        for (int i = 0; i < ArmorModule.MAX_PLATE_META; i++) {
-            subItems.add(ArmorModule.plateStack(i));
+    public void getSubItems(CreativeTabs tab, NonNullList<ItemStack> items) {
+        if(tab == EsteemedInnovation.tab) {
```
See ItemMetalNugget

elifoster commented on src/main/java/eiteam/esteemedinnovation/armor/exosuit/steam/ItemSteamExosuitArmor.java:
```diff
@@ -505,45 +507,44 @@ public boolean drainSteam(@Nonnull ItemStack me, int amountToDrain, EntityLiving
             if (hasPlates(me) && !"Thaumium".equals(UtilPlates.getPlate(me.getTagCompound().getString("Plate")).getIdentifier()) &&
-              !"Terrasteel".equals(UtilPlates.getPlate(me.getTagCompound().getString("Plate")).getIdentifier())) {
-                list.add(TextFormatting.BLUE + UtilPlates.getPlate(me.getTagCompound().getString("Plate")).effect());
+                    !"Terrasteel".equals(UtilPlates.getPlate(me.getTagCompound().getString("Plate")).getIdentifier())) {
```
This should be a 2 space indent, not 8, as per the code style

elifoster commented on src/main/java/eiteam/esteemedinnovation/armor/exosuit/steam/upgrades/ItemModularAcceleratorUpgrade.java:
```diff
@@ -32,7 +32,7 @@ public void onPlayerUpdate(LivingEvent.LivingUpdateEvent event, EntityPlayer pla
             if ((player.moveForward > 0.0F) && (lastX != player.posX || lastZ != player.posZ) && player.onGround && !player.isInWater()) {
                 ItemStack chestStack = player.getItemStackFromSlot(EntityEquipmentSlot.CHEST);
-                player.moveRelative(0F, 1F, 0.075F); //entity.moveFlying(0.0F, 1.0F, 0.075F); TODO Test this.
+                player.moveRelative(0F, 1F, 0.075F, 0.0F); //entity.moveFlying(0.0F, 1.0F, 0.075F); TODO Test this.
```
This should be shortened to `0F` for code style purposes

elifoster commented on src/main/java/eiteam/esteemedinnovation/commons/EsteemedInnovation.java:
```diff
@@ -88,15 +88,17 @@
     public static final String MISC_SECTION = "section.Misc.name";
     public static final String STEAMTOOL_SECTION = "section.SteamTools.name";
 
+    //TODO: Fix sound registering
```
What is the issue that needs fixing?

ExpensiveKoala replied:
`SoundEvent`s, like `Block`s and `Item`s, need to be registered on their own registry event.

elifoster commented on src/main/java/eiteam/esteemedinnovation/commons/handler/GenericTickHandler.java:
```diff
@@ -51,7 +50,7 @@
         if (FMLCommonHandler.instance().getSide() == Side.CLIENT) {
-            FMLLog.info("[EI] Getting some fields from reflection for Tick Handling.");
+            EsteemedInnovation.logger.info("[EI] Getting some fields from reflection for Tick Handling.");
             itemInMainHandField = ReflectionHelper.getField("itemStackMainHand", "field_187467_d", ItemRenderer.class);
```
Note (mostly to myself):
* Double check that these fields are correct
* Go through all ReflectionHelper usages

ExpensiveKoala committed:
Fix formatting

elifoster commented on build.gradle:
```diff
@@ -116,15 +116,36 @@ repositories {
+	maven {
+		name = "Jared"
+		url = "http://maven.blamejared.com/"
+	}
```
These lines are indented using the tab character. Can you change them to be 4 spaces instead?

elifoster commented on build.gradle:
```diff
@@ -116,15 +116,36 @@ repositories {
+	maven {
+        name = "JitPack"
+        url = "https://jitpack.io"
+    }
```
This maven block is indented using the wrong number of spaces

elifoster commented on build.gradle:
```diff
@@ -116,15 +116,36 @@ repositories {
+	deobfCompile enchiridionVersion
+	compile craftTweaker2APIVersion
+	compile craftTweaker2Version
+	deobfCompile baublesVersion
```
These lines use tabs instead of spaces

elifoster commented on build.gradle:
```diff
@@ -116,15 +116,36 @@ repositories {
-    compile ic2Version
-    compile jeiVersion
+    // compile ic2Version
+    runtime jeiVersion
```
Can you explain this?

ExpensiveKoala replied:
I don't remember my exact reasoning for changing it, but it doesn't really change anything by having it be compile vs runtime. Reverting this change as it shouldn't have been changed in the first place.

elifoster commented on src/main/java/eiteam/esteemedinnovation/commons/init/ContentModule.java:
```diff
@@ -93,66 +117,67 @@ protected Item setup(Item startingItem, String path) {
+
+    protected void setupItemBlock(RegistryEvent.Register<Item> event, Block startingBlock, CreativeTabs tab, Function<Block, ItemBlock> itemBlock) {
+        ItemBlock item = itemBlock.apply(startingBlock);
+        if(tab != null) {
```
Missing space before (

elifoster commented on src/main/java/eiteam/esteemedinnovation/hammer/ContainerSteamAnvil.java:
```diff
@@ -163,7 +163,7 @@ public void updateRepairOutput() {
 
             if (itemstack2 != null) {
-                flag = itemstack2.getItem() == Items.ENCHANTED_BOOK && Items.ENCHANTED_BOOK.getEnchantments(itemstack2).tagCount() > 0;
+                flag = itemstack2.getItem() == Items.ENCHANTED_BOOK && !EnchantmentHelper.getEnchantments(itemstack2).isEmpty();
```
Hm. It might be worth the readability to add a `hasEnchantments` function to `ItemStackUtility`. Not expecting that for this PR.

elifoster commented on src/main/java/eiteam/esteemedinnovation/metalcasting/mold/ItemMold.java:
```diff
@@ -30,10 +30,11 @@ public int getCostToMold(CrucibleLiquid liquid, @Nonnull ItemStack moldStack) {
-        for (Type type : Type.LOOKUP) {
-            subItems.add(new ItemStack(this, 1, type.ordinal()));
+    public void getSubItems(CreativeTabs tab, NonNullList<ItemStack> items) {
+        if(tab == EsteemedInnovation.tab) {
```
Missing space before (

elifoster commented:
I'm wondering, too, what the point of those `tab == EsteemedInnovation.tab` checks is. Why do we need that now?

ExpensiveKoala committed:
Fix formatting & use isInCreativeTab

ExpensiveKoala commented:
The reason for the checks in getSubItems is to prevent the items from showing up in every single tab. (Some change that 1.12 brought).

A good example of an issue caused by not having this is https://github.com/McJty/TheOneProbe/issues/169

ExpensiveKoala committed:
revert compile->runtime change

TehNut commented:
I'll just mention that you should be using `isInCreativeTab()` instead of the check you have now.

ExpensiveKoala commented:
Yeah I fixed that on the latest commit already

elifoster commented:
There's a few issues with how this is set up that are causing warnings and errors.

OreDictionary invalid registration attempts. It appears to be on every single entry. Here is an example warning:
```
[21:32:42] [main/WARN] [FML]: ****************************************
[21:32:42] [main/WARN] [FML]: * Invalid registration attempt for an Ore Dictionary item with name blockCopper has occurred. The registration has been denied to prevent crashes. The mod responsible for the registration needs to correct this.
[21:32:42] [main/WARN] [FML]: *  at net.minecraftforge.oredict.OreDictionary.registerOreImpl(OreDictionary.java:642)
[21:32:42] [main/WARN] [FML]: *  at net.minecraftforge.oredict.OreDictionary.registerOre(OreDictionary.java:628)
[21:32:42] [main/WARN] [FML]: *  at eiteam.esteemedinnovation.materials.MaterialsModule.oreDict(MaterialsModule.java:137)
[21:32:42] [main/WARN] [FML]: *  at eiteam.esteemedinnovation.commons.init.ContentModuleHandler.preInit(ContentModuleHandler.java:80)
[21:32:42] [main/WARN] [FML]: *  at eiteam.esteemedinnovation.commons.EsteemedInnovation.preInit(EsteemedInnovation.java:118)
[21:32:42] [main/WARN] [FML]: *  at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)...
[21:32:42] [main/WARN] [FML]: ****************************************
```

NullPointerException in setupItemBlock:
```
---- Minecraft Crash Report ----
// Shall we play a game?

Time: 2/13/18 9:32 PM
Description: Initializing game

java.lang.NullPointerException: Initializing game
	at eiteam.esteemedinnovation.commons.init.ContentModule.setupItemBlock(ContentModule.java:177)
	at eiteam.esteemedinnovation.commons.init.ContentModule.setupItemBlock(ContentModule.java:163)
	at eiteam.esteemedinnovation.materials.MaterialsModule.registerItems(MaterialsModule.java:126)
	at eiteam.esteemedinnovation.commons.init.ContentModuleHandler.registerItems(ContentModuleHandler.java:117)
	at eiteam.esteemedinnovation.commons.CommonProxy.registerItems(CommonProxy.java:41)
	at net.minecraftforge.fml.common.eventhandler.ASMEventHandler_8_CommonProxy_registerItems_Register.invoke(.dynamic)
	at net.minecraftforge.fml.common.eventhandler.ASMEventHandler.invoke(ASMEventHandler.java:90)
	at net.minecraftforge.fml.common.eventhandler.EventBus$1.invoke(EventBus.java:143)
	at net.minecraftforge.fml.common.eventhandler.EventBus.post(EventBus.java:179)
	at net.minecraftforge.registries.GameData.fireRegistryEvents(GameData.java:741)
	at net.minecraftforge.fml.common.Loader.preinitializeMods(Loader.java:604)
	at net.minecraftforge.fml.client.FMLClientHandler.beginMinecraftLoading(FMLClientHandler.java:270)
	at net.minecraft.client.Minecraft.init(Minecraft.java:513)
	at net.minecraft.client.Minecraft.run(Minecraft.java:421)
	at net.minecraft.client.main.Main.main(Main.java:118)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at net.minecraft.launchwrapper.Launch.launch(Launch.java:135)
	at net.minecraft.launchwrapper.Launch.main(Launch.java:28)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at net.minecraftforge.gradle.GradleStartCommon.launch(GradleStartCommon.java:97)
	at GradleStart.main(GradleStart.java:26)


A detailed walkthrough of the error, its code path and all known details is as follows:
---------------------------------------------------------------------------------------

-- Head --
Thread: Client thread
Stacktrace:
	at eiteam.esteemedinnovation.commons.init.ContentModule.setupItemBlock(ContentModule.java:177)
	at eiteam.esteemedinnovation.commons.init.ContentModule.setupItemBlock(ContentModule.java:163)
	at eiteam.esteemedinnovation.materials.MaterialsModule.registerItems(MaterialsModule.java:126)
	at eiteam.esteemedinnovation.commons.init.ContentModuleHandler.registerItems(ContentModuleHandler.java:117)
	at eiteam.esteemedinnovation.commons.CommonProxy.registerItems(CommonProxy.java:41)
	at net.minecraftforge.fml.common.eventhandler.ASMEventHandler_8_CommonProxy_registerItems_Register.invoke(.dynamic)
	at net.minecraftforge.fml.common.eventhandler.ASMEventHandler.invoke(ASMEventHandler.java:90)
	at net.minecraftforge.fml.common.eventhandler.EventBus$1.invoke(EventBus.java:143)
	at net.minecraftforge.fml.common.eventhandler.EventBus.post(EventBus.java:179)
	at net.minecraftforge.registries.GameData.fireRegistryEvents(GameData.java:741)
	at net.minecraftforge.fml.common.Loader.preinitializeMods(Loader.java:604)
	at net.minecraftforge.fml.client.FMLClientHandler.beginMinecraftLoading(FMLClientHandler.java:270)
	at net.minecraft.client.Minecraft.init(Minecraft.java:513)

-- Initialization --
Details:
Stacktrace:
	at net.minecraft.client.Minecraft.run(Minecraft.java:421)
	at net.minecraft.client.main.Main.main(Main.java:118)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at net.minecraft.launchwrapper.Launch.launch(Launch.java:135)
	at net.minecraft.launchwrapper.Launch.main(Launch.java:28)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at net.minecraftforge.gradle.GradleStartCommon.launch(GradleStartCommon.java:97)
	at GradleStart.main(GradleStart.java:26)

-- System Details --
Details:
	Minecraft Version: 1.12.2
	Operating System: Mac OS X (x86_64) version 10.13.1
	Java Version: 1.8.0_60, Oracle Corporation
	Java VM Version: Java HotSpot(TM) 64-Bit Server VM (mixed mode), Oracle Corporation
	Memory: 273516936 bytes (260 MB) / 897056768 bytes (855 MB) up to 1908932608 bytes (1820 MB)
	JVM Flags: 0 total; 
	IntCache: cache: 0, tcache: 0, allocated: 0, tallocated: 0
	FML: MCP 9.42 Powered by Forge 14.23.1.2604 17 mods loaded, 17 mods active
	States: 'U' = Unloaded 'L' = Loaded 'C' = Constructed 'H' = Pre-initialized 'I' = Initialized 'J' = Post-initialized 'A' = Available 'D' = Disabled 'E' = Errored

	| State | ID                    | Version         | Source                                       | Signature |
	|:----- |:--------------------- |:--------------- |:-------------------------------------------- |:--------- |
	| UCH   | minecraft             | 1.12.2          | minecraft.jar                                | None      |
	| UCH   | mcp                   | 9.42            | minecraft.jar                                | None      |
	| UCH   | FML                   | 8.0.99.99       | forgeSrc-1.12.2-14.23.1.2604.jar             | None      |
	| UCH   | forge                 | 14.23.1.2604    | forgeSrc-1.12.2-14.23.1.2604.jar             | None      |
	| UCH   | esteemedinnovationapi | @APIVERSION@    | Esteemed-Innovation_api                      | None      |
	| UCH   | codechickenlib        | 3.1.5.334       | CodeChickenLib-1.12.2-3.1.5.334-deobf.jar    | None      |
	| UCH   | baubles               | 1.5.2           | Baubles-573faf2.jar                          | None      |
	| UCH   | esteemedinnovation    | @VERSION@       | Esteemed-Innovation_main                     | None      |
	| UCH   | mantle                | 1.12-1.3.1.22   | Mantle-1.12-1.3.1.22.jar                     | None      |
	| UCH   | jei                   | 4.8.5.138       | jei_1.12.2-4.8.5.138.jar                     | None      |
	| UCH   | tconstruct            | 1.12.2-2.9.0.58 | TConstruct-1.12.2-2.9.0.58.jar               | None      |
	| UCH   | natura                | 1.12-4.3.0.8    | Natura-1.12-1.12-4.3.0.8.jar                 | None      |
	| UCH   | crafttweaker          | 4.1.0           | CraftTweaker2-MC1120-Main-1.12-4.1.0.395.jar | None      |
	| UCH   | ctgui                 | 1.0.0           | CraftTweaker2-MC1120-Main-1.12-4.1.0.395.jar | None      |
	| UCH   | guideapi              | 1.12-2.1.4-57   | Guide-API-1.12-2.1.4-57.jar                  | None      |
	| UCH   | enchiridion           | 3.3.3           | enchiridion_1.12.2-1.12.2-3.3.3-4.jar        | None      |
	| UCH   | theoneprobe           | 1.4.19          | theoneprobe-1.12-1.4.19.jar                  | None      |

	Loaded coremods (and transformers): 
	Pulsar/tconstruct loaded Pulses: 
		- TinkerCommons (Enabled/Forced)
		- TinkerWorld (Enabled/Not Forced)
		- TinkerTools (Enabled/Not Forced)
		- TinkerHarvestTools (Enabled/Forced)
		- TinkerMeleeWeapons (Enabled/Forced)
		- TinkerRangedWeapons (Enabled/Forced)
		- TinkerModifiers (Enabled/Forced)
		- TinkerSmeltery (Enabled/Not Forced)
		- TinkerGadgets (Enabled/Not Forced)
		- TinkerOredict (Enabled/Forced)
		- TinkerIntegration (Enabled/Forced)
		- TinkerFluids (Enabled/Forced)
		- TinkerMaterials (Enabled/Forced)
		- TinkerModelRegister (Enabled/Forced)
		- theoneprobeIntegration (Enabled/Not Forced)

	Pulsar/natura loaded Pulses: 
		- NaturaCommons (Enabled/Forced)
		- NaturaOverworld (Enabled/Not Forced)
		- NaturaNether (Enabled/Not Forced)
		- NaturaDecorative (Enabled/Not Forced)
		- NaturaTools (Enabled/Not Forced)
		- NaturaEntities (Enabled/Not Forced)
		- NaturaOredict (Enabled/Forced)
		- NaturaWorld (Enabled/Not Forced)

	Launched Version: 1.12.2
	LWJGL: 2.9.2
	OpenGL: NVIDIA GeForce GT 330M OpenGL Engine GL version 2.1 NVIDIA-10.4.12 310.90.30.05b27, NVIDIA Corporation
	GL Caps: Using GL 1.3 multitexturing.
Using GL 1.3 texture combiners.
Using framebuffer objects because ARB_framebuffer_object is supported and separate blending is supported.
Shaders are available because OpenGL 2.1 is supported.
VBOs are available because OpenGL 1.5 is supported.

	Using VBOs: Yes
	Is Modded: Definitely; Client brand changed to 'fml,forge'
	Type: Client (map_client.txt)
	Resource Packs: 
	Current Language: English (US)
	Profiler Position: N/A (disabled)
	CPU: 4x Intel(R) Core(TM) i7 CPU M 620 @ 2.67GHz
```

I'm not so concerned with the OreDict warnings, since you did not touch the OreDict. I can resolve those after merging if you don't feel like messing with it, but if you could resolve this ItemBlock issue in the implementation you've provided in this PR that would be excellent.

ExpensiveKoala committed:
Add few missing things. Move oredict to init.

Registration now happens between preinit and init so we can only register oredict at init or later. Ideally we want to register oredict inside the item registration event.

ExpensiveKoala committed:
Add modid param to EventBusSubscribers

This fixes the oredict issue.

elifoster commented on src/api/java/eiteam/esteemedinnovation/api/recipe/ShapedRecipeFactory.java:
```diff
@@ -32,7 +32,10 @@ public IRecipe parse(JsonContext context, JsonObject json) {
 
-        BookRecipeRegistry.addRecipe(JsonUtils.getString(json, "book_key"), shapedRecipe);
+        String bookKey = JsonUtils.getString(json, "book_key", "");
+        if (!bookKey.equals("")) {
```
`!bookKey.equals("")` -> `!bookKey.isEmpty()`

elifoster commented on src/api/java/eiteam/esteemedinnovation/api/recipe/ShapelessRecipeFactory.java:
```diff
@@ -39,7 +39,10 @@ public IRecipe parse(JsonContext context, JsonObject json) {
         ShapelessRecipe shapelessRecipe = new ShapelessRecipe(group.isEmpty() ? null : new ResourceLocation(group), ings, itemstack);
-        BookRecipeRegistry.addRecipe(JsonUtils.getString(json, "book_key"), shapelessRecipe);
+        String bookKey = JsonUtils.getString(json, "book_key", "");
+        if (!bookKey.equals("")) {
```
`!bookKey.equals("")` -> `!bookKey.isEmpty()`

elifoster commented on src/main/resources/assets/esteemedinnovation/recipes/astrolabe.json:
```diff
@@ -1,6 +1,7 @@
 {
   "type": "esteemedinnovationapi:shaped_recipe",
   "group": "wooden_fence",
+  "book_key": "woodenCone",
```
Was this intentional, and if so can you explain your reasoning?

elifoster commented on src/main/java/eiteam/esteemedinnovation/tools/ToolsModule.java:
```diff
@@ -141,6 +141,8 @@ public void registerItems(RegistryEvent.Register<Item> event) {
         SIFTER = setupUpgrade(event, new ItemSifterUpgrade(), "sifter");
         BACKHOE = setupUpgrade(event, new ItemBackhoeUpgrade(), "backhoe");
+
+        DrillHeadMaterial.registerDefaults();
```
Why was this moved? This must be done after all mod items and OreDict entries (not just EI) have been registered. See: https://github.com/Esteemed-Innovation/Esteemed-Innovation/blob/4230f9e3a89f9175afc720fcc6575300058d16b3/src/main/java/eiteam/esteemedinnovation/tools/steam/upgrades/drillhead/DrillHeadMaterial.java#L126

ExpensiveKoala replied:
Because it was in the Recipes method. Recipes happens before item registration so it needed to be moved.

elifoster replied:
Ah, I was unaware that the time recipes are set up changed. Hm. It might be better to move this into `ToolsModule#finish()`

elifoster commented on src/main/resources/assets/esteemedinnovation/recipes/astrolabe.json:
```diff
@@ -0,0 +1,30 @@
+{
+  "type": "esteemedinnovationapi:shaped_recipe",
+  "group": "wooden_fence",
+  "book_key": "woodenCone",
```
Was this intentional, and if so can you explain your reasoning?

elifoster commented on src/api/java/eiteam/esteemedinnovation/api/recipe/ShapelessRecipeFactory.java:
```diff
@@ -0,0 +1,55 @@
+
+        ShapelessRecipe shapelessRecipe = new ShapelessRecipe(group.isEmpty() ? null : new ResourceLocation(group), ings, itemstack);
+        String bookKey = JsonUtils.getString(json, "book_key", "");
+        if (!bookKey.equals("")) {
```
`!bookKey.equals("")` -> `!bookKey.isEmpty()`

elifoster commented on src/api/java/eiteam/esteemedinnovation/api/recipe/ShapedRecipeFactory.java:
```diff
@@ -0,0 +1,48 @@
+        ShapedRecipe shapedRecipe = new ShapedRecipe(group.isEmpty() ? null : new ResourceLocation(group), recipe.getRecipeOutput(), primer);
+
+        String bookKey = JsonUtils.getString(json, "book_key", "");
+        if (!bookKey.equals("")) {
```
`!bookKey.equals("")` -> `!bookKey.isEmpty()`

ExpensiveKoala committed:
Fix json book_key

I don't know how I ended up getting woodenCone

ExpensiveKoala committed:
string isEmpty check

elifoster commented:
Do you think it would be better to have `ClientProxy.registerModels(ModelRegistryEvent)`, `CommonProxy.registerBlocks(RegistryEvent.Register<Block>)`, and `CommonProxy.registerItems(RegistryEvent.Register<Item>)` inside of `ContentModuleHandler` directly as opposed to `ClientProxy` and `CommonProxy`, or is there a reason these _must_ be inside the SidedProxy classes?

ExpensiveKoala committed:
Move DrillHeadMaterial.registerDefaults() to finish

ExpensiveKoala commented:
There isn't really. Do you want me to move it?

elifoster commented:
I think that would make sense. It would keep all of that stuff inside the same class.

ExpensiveKoala committed:
Move registry events to ContentModule Handler.

Also null check on ItemDrillHeadUpgrade

elifoster committed:
:lipstick: Remove unnecessary this qualifiers

elifoster commented on src/main/java/eiteam/esteemedinnovation/materials/raw/BlockGenericOre.java:
```diff
@@ -48,10 +49,11 @@ public IBlockState getStateFromMeta(int meta) {
-        for (OreBlockTypes type : OreBlockTypes.LOOKUP) {
-            list.add(new ItemStack(this, 1, type.getMetadata()));
+    public void getSubBlocks(CreativeTabs tab, NonNullList<ItemStack> items) {
+        if (tab == EsteemedInnovation.tab) {
```
This should be `isInCreativeTab`

elifoster commented on src/main/java/eiteam/esteemedinnovation/materials/refined/BlockBeacon.java:
```diff
@@ -33,9 +34,11 @@ public boolean isBeaconBase(IBlockAccess worldObj, BlockPos pos, BlockPos beacon
-        for (MetalBlockTypes type : MetalBlockTypes.values()) {
-            list.add(new ItemStack(this, 1, type.getMetadata()));
+    public void getSubBlocks(CreativeTabs tab, NonNullList<ItemStack> items) {
+        if (tab == EsteemedInnovation.tab) {
```
This should be `isInCreativeTab`

elifoster committed:
:lipstick: Minor spacing corrections

elifoster committed:
:package: New util function EnchantmentUtility.hasEnchantments(ItemStack)

elifoster committed:
:lipstick: Replace config category strings with constants

elifoster committed:
:gem: :package: Implement new system for configuration in ContentModules
New ConfigurableModule interface to be added onto ContentModule child classes so that they can have their own configuration options within the EsteemedInnovation.cfg config file.

elifoster committed:
:package: Reimplement Thumper config options to use new ConfigurableModule system

elifoster committed:
:package: Reimplement Smasher config options to use new ConfigurableModule system

elifoster committed:
:package: Reimplement Buzzsaw config options to use new ConfigurableModule system

elifoster committed:
:package: Reimplement Charging config options to use new ConfigurableModule system

elifoster committed:
:package: Reimplement Converter config options to use new ConfigurableModule system

elifoster committed:
:package: Reimplement Hammer config options to use new ConfigurableModule system

elifoster committed:
:package: Reimplement Heater config options to use new ConfigurableModule system

elifoster committed:
:package: Reimplement Metalcasting config options to use new ConfigurableModule system

elifoster committed:
:bug: Add missing return statement in MetalcastingModule#isRecipeEnabled(String)

elifoster committed:
:package: Reimplement Boiler configs using ConfigurableModule

elifoster committed:
:package: Reimplement Engineering Table configs using ConfigurableModule

elifoster committed:
:package: Reimplement Firearm configs using ConfigurableModule

elifoster committed:
:package: :bug: Move disableMainBarrelRecipe to MiscellaneousModule

elifoster committed:
:package: Reimplement Steam Safety configs using ConfigurableModule

elifoster committed:
:package: Reimplement Storage configs using ConfigurableModule

elifoster committed:
:package: Reimplement Wooden Code configs using ConfigurableModule

elifoster committed:
:package: Reimplement Workshop configs using ConfigurableModule

elifoster committed:
:package: Reimplement Transport configs using ConfigurableModule

elifoster committed:
:package: Reimplement Armor configs using ConfigurableModule

elifoster committed:
:package: Reimplement Tool configs using ConfigurableModule

elifoster committed:
:lipstick: :package: Clean up Config, move some leftover fields to their proper modules
The remaining ones are either general or are not obviously used elsewhere.

ExpensiveKoala committed:
Fix astrolabe recipe

elifoster committed:
:fire: Remove our Iron Nugget
Vanilla Minecraft now includes an Iron Nugget

elifoster committed:
:package: :book: Update removeRecipe code for new recipe registry
Update to use the new Stream API in Java 8
Add a new removeRecipe overload that takes a ResourceLocation directly
Clean up unnecessary variable declarations and instantiations and null checks for now-nonnull objects
Update Hopper recipe removal to use removeRecipeByOutput
Add and update javadocs

elifoster committed:
:bug: Add missing book_key to book recipe JSON, remove TODO from BookModule

ExpensiveKoala committed:
📦Add recipe methods to ContentModule

elifoster commented on src/api/java/eiteam/esteemedinnovation/api/book/BookRecipeRegistry.java:
```diff
@@ -11,12 +12,12 @@
 
     public static void addRecipe(String key, ItemStack output, Object... params) {
-        ShapedOreRecipe recipe = new ShapedOreRecipe(output, params);
+        ResourceLocation group = new ResourceLocation(Constants.EI_MODID, key);
```
Because this is in the API, it is for use beyond just EI, so this method should probably be `@Deprecated` and then an overload should be made which takes a `ResourceLocation` instead of just the `String` key.

```java
/**
 * @deprecated Use {@link #addRecipe(ResourceLocation, ItemStack, Object...} instead
 */
@Deprecated
public static void addRecipe(String key, ItemStack output, Object... params) {
    addRecipe(new ResourceLocation(Constants.EI_MODID, key), output, params);
}

public static void addRecipe(ResourceLocation group, ItemStack output, Object... params) {
    ShapedOreRecipe recipe = new ShapedOreRecipe(group, output, params);
    addRecipe(group.getName()/* I can't remember the exact name of the function right now */, recipe);
}
```

If we want a non-deprecated helper method that defaults to EI_MODID we can add one in the mod itself rather than the API.

ExpensiveKoala committed:
📦Deprecate old addRecipe

ExpensiveKoala committed:
📦Re-add Boiler Module recipes

ExpensiveKoala committed:
📦Re-add Armor Module Recipes

ExpensiveKoala committed:
📦Re-add Buzzsaw Module Recipes

ExpensiveKoala committed:
💄Clean up Boiler Module indentation

ExpensiveKoala committed:
📦Re-add Charging Module recipes

ExpensiveKoala committed:
📦Re-add Converter Module recipes

ExpensiveKoala committed:
📦Re-add Engineering Table Module recipes

ExpensiveKoala committed:
📦Re-add Firearm Module recipes

ExpensiveKoala committed:
📦Re-add Hammer Module recipes

ExpensiveKoala committed:
📦Re-add Heater Module recipes

ExpensiveKoala committed:
📦Re-add Materials Module recipes

ExpensiveKoala committed:
📦Re-add Metalcasting Module recipes

ExpensiveKoala committed:
📦Re-add Miscellaneous Module recipes

ExpensiveKoala committed:
📦Re-add Natural Philosophy Module Recipe

ExpensiveKoala committed:
📦Re-add Pendulum Module recipes

ExpensiveKoala committed:
📦Re-add Smasher Module recipes

ExpensiveKoala committed:
📦Re-add Safety Module recipes

ExpensiveKoala committed:
📦Re-add Storage Module recipes

ExpensiveKoala committed:
📦Re-add Thumper Module recipes

ExpensiveKoala committed:
📦Re-add Wooden Cone Module recipes

ExpensiveKoala committed:
📦Steam Workshop Module fix method header

ExpensiveKoala committed:
📦Re-add Transportation Module recipes

ExpensiveKoala committed:
📦Re-add Tools Module recipes

ExpensiveKoala committed:
🐛Fix various recipe related crashes

elifoster committed:
:book: :shirt: Improve ContentModule docs, remove deprecated ContentModule#recipes(Side) and ContentModule#oreDict(Side) APIs

elifoster committed:
:broken_heart: Remove deprecated BookRecipeRegistry.addRecipe(String, ItemStack, Object...) method

elifoster committed:
:package: :book: Move recipe adding methods to RecipeUtility, add documentation for them

ExpensiveKoala committed:
🐛Fix Natura maven location

ExpensiveKoala committed:
🐛Fix Steam Tool stack null-pointer

ExpensiveKoala committed:
🐛Fix firearm->engineering table crash

ExpensiveKoala committed:
🐛Fix Engineering table not being dropped by drill

Depending on how the block is implemented, the drill might have problems with it.

ExpensiveKoala committed:
🐛Fix various item rendering being broken

ExpensiveKoala committed:
🐛Fix exosuit model not showing

elifoster committed:
🐛 Fix Leather Exosuit item model not appearing

ExpensiveKoala committed:
🐛Fix broken Steam Exosuit rendering

elifoster committed:
💔 Remove ConfigurableModule recipe-related methods and recipe JSONs

elifoster committed:
🐛 Fix missing book recipe

ExpensiveKoala committed:
🐛Fix Steam Tools bobbing when rightclicking

elifoster commented on src/api/java/eiteam/esteemedinnovation/api/recipe/ShapedRecipeFactory.java:
```diff
@@ -0,0 +1,48 @@
+import net.minecraftforge.common.crafting.JsonContext;
+import net.minecraftforge.oredict.ShapedOreRecipe;
+
+public class ShapedRecipeFactory implements IRecipeFactory {
```
This class should be removed, I believe.

elifoster commented on src/api/java/eiteam/esteemedinnovation/api/recipe/ShapelessRecipeFactory.java:
```diff
@@ -0,0 +1,55 @@
+import net.minecraftforge.common.crafting.JsonContext;
+import net.minecraftforge.oredict.ShapelessOreRecipe;
+
+public class ShapelessRecipeFactory implements IRecipeFactory {
```
This class should be removed, I believe.

elifoster commented on src/api/java/eiteam/esteemedinnovation/api/research/ResearchRecipe.java:
```diff
@@ -36,7 +38,7 @@
      */
     public ResearchRecipe(@Nonnull ItemStack result, TriPredicate<InventoryCrafting, EntityPlayer, World> matcher, Object... recipe) {
-        super(result, recipe);
+        super(new ResourceLocation(Constants.EI_MODID, "shaped_research"), result, recipe);
```
This should use the API mod ID since this is an API feature.

elifoster commented on src/api/java/eiteam/esteemedinnovation/api/research/ShapelessResearchRecipe.java:
```diff
@@ -27,7 +29,7 @@
      */
     public ShapelessResearchRecipe(@Nonnull ItemStack result, TriPredicate<InventoryCrafting, EntityPlayer, World> matcher, Object... recipe) {
-        super(result, recipe);
+        super(new ResourceLocation(Constants.EI_MODID, "shapeless_research"), result, recipe);
```
This should use the API mod ID since this is an API feature.

elifoster commented on src/api/resources/assets/esteemedinnovationapi/recipes/_factories.json:
```diff
@@ -0,0 +1,6 @@
+{
```
This file can be removed, I believe.

elifoster commented on src/main/java/eiteam/esteemedinnovation/armor/ArmorModule.java:
```diff
@@ -166,44 +189,117 @@
-        DRAGON_ROAR = setup(new ItemDragonRoarUpgrade(), "dragon_roar");
-        EXOSUIT_PLATE = setup(new ItemExosuitPlate(), "exosuit_plate", EsteemedInnovation.tab, false);
-        PISTON_PUSH = setup(new ItemPistonPushUpgrade(), "piston_push");
+
```
Unnecessary blank line here

elifoster commented on src/main/java/eiteam/esteemedinnovation/armor/ArmorModule.java:
```diff
@@ -260,552 +351,550 @@ public static String resource(@Nonnull String base) {
-              't', SPYGLASS,
-              'g', BLOCK_GLASS_COLORLESS
-            ));
+    public void recipes(RegistryEvent.Register<IRecipe> event) {
```
The continuation indentations in this method implementation use 8 space tabs instead of 2 space tabs.

elifoster commented on src/main/java/eiteam/esteemedinnovation/armor/exosuit/steam/ItemSteamExosuitArmor.java:
```diff
@@ -505,45 +506,44 @@ public boolean drainSteam(@Nonnull ItemStack me, int amountToDrain, EntityLiving
             if (hasPlates(me) && !"Thaumium".equals(UtilPlates.getPlate(me.getTagCompound().getString("Plate")).getIdentifier()) &&
-              !"Terrasteel".equals(UtilPlates.getPlate(me.getTagCompound().getString("Plate")).getIdentifier())) {
-                list.add(TextFormatting.BLUE + UtilPlates.getPlate(me.getTagCompound().getString("Plate")).effect());
+                  !"Terrasteel".equals(UtilPlates.getPlate(me.getTagCompound().getString("Plate")).getIdentifier())) {
```
This continuation indent uses 8 space tabs instead of 2 space tabs

elifoster commented on src/main/java/eiteam/esteemedinnovation/buzzsaw/BuzzsawModule.java:
```diff
@@ -1,40 +1,57 @@
+    @Override
+    public void recipes(RegistryEvent.Register<IRecipe> event) {
+        if (enableSaw) {
+            RecipeUtility.addRecipe(event, true, "saw", BUZZSAW,
```
This continuation indent uses 8 space tabs instead of 2 space tabs

elifoster commented on src/main/java/eiteam/esteemedinnovation/commons/init/ContentModule.java:
```diff
@@ -93,66 +122,66 @@ protected Item setup(Item startingItem, String path) {
 
     /**
-     * Overload for {@link #setup(Block, String, Function)} that uses the default constructor for {@link ItemBlock} as
+     * Overload for {@link #setup(RegistryEvent.Register, Block, String)} that uses the default constructor for {@link ItemBlock} as
```
This documentation is incorrect.

elifoster commented on src/main/java/eiteam/esteemedinnovation/commons/init/ContentModule.java:
```diff
@@ -93,66 +122,66 @@ protected Item setup(Item startingItem, String path) {
         return startingBlock;
     }
 
+    protected void setupItemBlock(RegistryEvent.Register<Item> event, Block startingBlock) {
```
This function should have documentation

elifoster commented on src/main/java/eiteam/esteemedinnovation/commons/init/ContentModule.java:
```diff
@@ -93,66 +122,66 @@ protected Item setup(Item startingItem, String path) {
+        setupItemBlock(event, startingBlock, EsteemedInnovation.tab);
+    }
+
+    protected void setupItemBlock(RegistryEvent.Register<Item> event, Block startingBlock, Function<Block, ItemBlock> itemBlock) {
```
This function should have documentation

elifoster commented on src/main/java/eiteam/esteemedinnovation/commons/init/ContentModule.java:
```diff
@@ -93,66 +122,66 @@ protected Item setup(Item startingItem, String path) {
+        setupItemBlock(event, startingBlock, EsteemedInnovation.tab, itemBlock);
+    }
+
+    protected void setupItemBlock(RegistryEvent.Register<Item> event, Block startingBlock, CreativeTabs tab) {
```
This function should have documentation

elifoster commented on src/main/java/eiteam/esteemedinnovation/commons/init/ContentModule.java:
```diff
@@ -93,66 +122,66 @@ protected Item setup(Item startingItem, String path) {
+        setupItemBlock(event, startingBlock, tab, ItemBlock::new);
+    }
+
+    protected void setupItemBlock(RegistryEvent.Register<Item> event, Block startingBlock, CreativeTabs tab, Function<Block, ItemBlock> itemBlock) {
```
This function should have documentation

elifoster commented on src/main/java/eiteam/esteemedinnovation/commons/init/ContentModule.java:
```diff
@@ -93,66 +122,66 @@ protected Item setup(Item startingItem, String path) {
+        if (tab != null) {
+            item.setCreativeTab(tab);
+        }
+        //item.setUnlocalizedName(Constants.EI_MODID + ":" + path);
```
This comment can probably be removed.

elifoster commented on src/main/java/eiteam/esteemedinnovation/commons/util/ReflectionHelper.java:
```diff
@@ -50,7 +50,7 @@ public static Method getMethod(String methodName, String obfName, Class clazz, C
             method = clazz.getDeclaredMethod(obfName, params);
         } catch (NoSuchMethodException e) {
-            FMLLog.warning("[EI] Unable to find method " + methodName + " with its obfuscated name. Trying to" +
+            EsteemedInnovation.logger.warn("[EI] Unable to find method " + methodName + " with its obfuscated name. Trying to" +
```
`[EI]` should be removed from this, because it says `[esteemedinnovation]` in each log entry

elifoster commented on src/main/java/eiteam/esteemedinnovation/commons/handler/GenericTickHandler.java:
```diff
@@ -51,7 +50,7 @@
     static {
         if (FMLCommonHandler.instance().getSide() == Side.CLIENT) {
-            FMLLog.info("[EI] Getting some fields from reflection for Tick Handling.");
+            EsteemedInnovation.logger.info("[EI] Getting some fields from reflection for Tick Handling.");
```
`[EI]` should be removed from this, because it says `[esteemedinnovation]` in each log entry

elifoster commented on src/main/java/eiteam/esteemedinnovation/commons/util/ReflectionHelper.java:
```diff
@@ -23,7 +23,7 @@ public static Field getField(String fieldName, String obfName, Class clazz) {
             field = clazz.getDeclaredField(obfName);
         } catch (NoSuchFieldException e) {
-            FMLLog.warning("[EI] Unable to find field " + fieldName + " with its obfuscated " +
+            EsteemedInnovation.logger.warn("[EI] Unable to find field " + fieldName + " with its obfuscated " +
```
`[EI]` should be removed from this, because it says `[esteemedinnovation]` in each log entry

elifoster commented on src/main/java/eiteam/esteemedinnovation/commons/util/ReflectionHelper.java:
```diff
@@ -65,7 +65,7 @@ public static Method getMethod(String methodName, String obfName, Class clazz, C
                     builderHasAMethod = true;
                 }
-                FMLLog.warning("[EI] Unable to find " + methodName + " method in " + clazz.getName() +
+                EsteemedInnovation.logger.warn("[EI] Unable to find " + methodName + " method in " + clazz.getName() +
```
`[EI]` should be removed from this, because it says `[esteemedinnovation]` in each log entry

elifoster commented on src/main/java/eiteam/esteemedinnovation/commons/util/ReflectionHelper.java:
```diff
@@ -85,7 +85,7 @@ private static void setAccessible(Method method) {
 
     public static void init() {
-        FMLLog.info("[EI] Getting some fields through reflection.");
+        EsteemedInnovation.logger.info("[EI] Getting some fields through reflection.");
```
`[EI]` should be removed from this, because it says `[esteemedinnovation]` in each log entry

elifoster commented on src/main/java/eiteam/esteemedinnovation/commons/util/ReflectionHelper.java:
```diff
@@ -102,7 +102,7 @@ public static void init() {
             setAccessible(merchantField);
         } catch (NoClassDefFoundError ignore) {
-            FMLLog.warning("[EI] GuiMerchant class not found. You are probably a server.");
+            EsteemedInnovation.logger.warn("[EI] GuiMerchant class not found. You are probably a server.");
```
`[EI]` should be removed from this, because it says `[esteemedinnovation]` in each log entry

elifoster commented on src/main/java/eiteam/esteemedinnovation/materials/raw/config/OreConfigurationParser.java:
```diff
@@ -246,7 +246,7 @@ private void writeDefault() throws IOException {
             } else {
                 if (pieces.length > 3) {
-                    FMLLog.warning("[EI] More than 3 values separated by : in ReplaceableBlocks '%s' for Ore generation config. Expected maximum: 'domain:path:meta'. This may indicate a bug in your config!", str);
+                    EsteemedInnovation.logger.warn("[EI] More than 3 values separated by : in ReplaceableBlocks '%s' for Ore generation config. Expected maximum: 'domain:path:meta'. This may indicate a bug in your config!", str);
```
`[EI]` should be removed from this, because it says `[esteemedinnovation]` in each log entry

elifoster commented on src/main/java/eiteam/esteemedinnovation/commons/Config.java:
```diff
@@ -58,400 +44,48 @@
+        enableRailcraftIntegration = config.get(CATEGORY_INTEGRATION, "Enable Railcraft", true).getBoolean();
+        enableNEIIntegration = config.get(CATEGORY_INTEGRATION, "Enable NEI", true).getBoolean();
+
+        // TODO: Abstract this somehow
```
@ExpensiveKoala Do you have any ideas for this? The one thing that stands out in my mind would be something like

```java
hasAllCrucial = true;
for (ConfigurableModule module : ContentModuleHandler.configurableModules) {
    if (!module.areCrucialOptionsEnabled()) {
        hasAllCrucial = false;
        break;
    }
}

// and in ConfigurableModule
default boolean areCrucialOptionsEnabled() {
    return true;
}

// example implementation in BoilerModule
@Override
public void areCrucialOptionsEnabled() {
    return enableBoiler;
}
```

ExpensiveKoala replied:
I suppose that would be better than what is currently there.

ExpensiveKoala committed:
💄💔Fix formatting, Add better JavaDocs, Remove unused Recipe factories

ExpensiveKoala committed:
📦Abstract crucial options

ExpensiveKoala committed:
💄Fix formatting of ArmorModule

elifoster commented on src/main/java/eiteam/esteemedinnovation/commons/init/ConfigurableModule.java:
```diff
@@ -14,4 +14,8 @@
      */
     void loadConfigurationOptions(Configuration config);
+
+    default boolean areCrucialOptionsEnabled() {
```
This needs documentation.

elifoster commented on src/main/java/eiteam/esteemedinnovation/commons/init/ContentModule.java:
```diff
@@ -27,20 +27,20 @@
  * <br />
```
Formatting in this class is not consistent with the code style. Notably:
* [The newlines after { for empty blocks](https://github.com/Esteemed-Innovation/Esteemed-Innovation/blob/1.12/STYLE_GUIDE.md#empty-blocks)

And things which are not explicitly defined in the code style document but are the style elsewhere in the code:
* Tabs before `<li>`s
* No newlines before `@param`s
* No unneeded whitespace after `@param`s

elifoster commented on src/main/java/eiteam/esteemedinnovation/armor/ArmorModule.java:
```diff
@@ -88,7 +88,7 @@
 import static net.minecraft.init.Items.*;
 
 public class ArmorModule extends ContentModule implements ConfigurableModule {
-    public static final ItemArmor.ArmorMaterial STEAM_EXO_MAT = EnumHelper.addArmorMaterial("STEAMEXOSUIT", MOD_ID + ":steam_exo", 15, new int[] { 2, 5, 4, 1 }, 0, SoundEvents.ITEM_ARMOR_EQUIP_GENERIC, 0F);
```
There should be a space between `[]` and `{`. The ones inside are optional as defined in the code style, but perhaps that should be changed to be one or the other.

elifoster commented on src/main/java/eiteam/esteemedinnovation/armor/ArmorModule.java:
```diff
@@ -889,9 +887,10 @@ private static void addFootRecipe(RegistryEvent.Register<IRecipe> event, Item it
     /**
      * Adds exosuit plate recipe (2x2 of ingredient), and a melting recipe for the itemstack.
-     * @param str The name of the recipe for the book.
+     *
```
The whitespace here and after the parameter names are unnecessary.

elifoster committed:
📖💄 Fix ContentModule code style and documentation
Add missing @param tags, because JavaDoc simply omits the parameters altogether if that is not specified, even if there is no description.
Remove unnecessary whitespace before and after @param tags.
Remove newlines in empty functions.

elifoster committed:
💔 Remove Config.hasAllCrucial field because it is not used anywhere

elifoster committed:
‼💄 Fix code style in ArmorModule
Remove whitespaces in and around @param tags.
Add whitespaces between [] and { in array initializers.
Fix unnecessarily qualified fields.
